// lwc-webpack-plugin/dist/loader.js
"use strict";
exports.__esModule = true;
var compiler = require('@lwc/compiler');
var babel = require('@babel/core');
var getInfoFromPath = require('./module').getInfoFromPath;
var loader_utils_1 = require("loader-utils");
module.exports = function (source) {
    var resourcePath = this.resourcePath;
    var _a = loader_utils_1.getOptions(this), stylesheetConfig = _a.stylesheetConfig, outputConfig = _a.outputConfig, experimentalDynamicComponent = _a.experimentalDynamicComponent;
    var info;
    try {
        info = getInfoFromPath(resourcePath, process.cwd());
    }
    catch (e) {
        info = {
            name: '',
            namespace: ''
        };
    }
    var codeTransformed = source;
    if (resourcePath.endsWith('.ts')) {
        var code = babel.transform(source, {
            filename: resourcePath,
            plugins: [
                require.resolve('@babel/plugin-syntax-class-properties'),
                [
                    require.resolve('@babel/plugin-syntax-decorators'),
                    {
                        decoratorsBeforeExport: true
                    }
                ],
                require.resolve('@babel/plugin-proposal-optional-chaining'),
                require.resolve('@babel/plugin-proposal-nullish-coalescing-operator')
            ],
            presets: [
                require.resolve('@babel/preset-typescript'),
                {
                    allowDeclareFields: true
                }
            ]
        }).code;
        codeTransformed = code;
    }
    var cb = this.async();
    compiler
        .transform(codeTransformed, resourcePath, {
        name: info.ns,
        namespace: 'my',
        stylesheetConfig: stylesheetConfig,
        outputConfig: outputConfig,
        experimentalDynamicComponent: experimentalDynamicComponent
    })
        .then(function (res) {
        cb(null, res.code);
    })["catch"](function (err) {
        cb(err);
    });
    return;
};
