// @lwc/compiler/dist/commonjs/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateTransformOptions = void 0;
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const errors_1 = require("@lwc/errors");
const shared_1 = require("@lwc/shared");
const DEFAULT_OPTIONS = {
    isExplicitImport: false,
    preserveHtmlComments: false,
};
const DEFAULT_DYNAMIC_CMP_CONFIG = {
    loader: '',
    strictSpecifier: true,
};
const DEFAULT_STYLESHEET_CONFIG = {
    customProperties: {
        resolution: { type: 'native' },
    },
};
const DEFAULT_OUTPUT_CONFIG = {
    minify: false,
    sourcemap: false,
};
function validateTransformOptions(options) {
    validateOptions(options);
    return normalizeOptions(options);
}
exports.validateTransformOptions = validateTransformOptions;
function validateOptions(options) {
    errors_1.invariant(!shared_1.isUndefined(options), errors_1.CompilerValidationErrors.MISSING_OPTIONS_OBJECT, [options]);
    if (!shared_1.isUndefined(options.stylesheetConfig)) {
        validateStylesheetConfig(options.stylesheetConfig);
    }
    if (!shared_1.isUndefined(options.outputConfig)) {
        validateOutputConfig(options.outputConfig);
    }
}
function validateStylesheetConfig(config) {
    const { customProperties } = config;
    if (!shared_1.isUndefined(customProperties)) {
        const { resolution } = customProperties;
        if (!shared_1.isUndefined(resolution)) {
            errors_1.invariant(shared_1.isObject(resolution), errors_1.CompilerValidationErrors.INVALID_RESOLUTION_PROPERTY, [
                resolution,
            ]);
            const { type } = resolution;
            errors_1.invariant(type === 'native' || type === 'module', errors_1.CompilerValidationErrors.INVALID_TYPE_PROPERTY, [type]);
        }
    }
}
function isUndefinedOrBoolean(property) {
    return shared_1.isUndefined(property) || shared_1.isBoolean(property);
}
function validateOutputConfig(config) {
    errors_1.invariant(isUndefinedOrBoolean(config.minify), errors_1.CompilerValidationErrors.INVALID_MINIFY_PROPERTY, [config.minify]);
    errors_1.invariant(isUndefinedOrBoolean(config.sourcemap), errors_1.CompilerValidationErrors.INVALID_SOURCEMAP_PROPERTY, [config.sourcemap]);
}
function normalizeOptions(options) {
    const outputConfig = {
        ...DEFAULT_OUTPUT_CONFIG,
        ...options.outputConfig,
    };
    const stylesheetConfig = {
        customProperties: {
            ...DEFAULT_STYLESHEET_CONFIG.customProperties,
            ...(options.stylesheetConfig && options.stylesheetConfig.customProperties),
        },
    };
    const experimentalDynamicComponent = {
        ...DEFAULT_DYNAMIC_CMP_CONFIG,
        ...options.experimentalDynamicComponent,
    };
    return {
        ...DEFAULT_OPTIONS,
        ...options,
        stylesheetConfig,
        outputConfig,
        experimentalDynamicComponent,
    };
}
