// @oclif/core/lib/help/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodashTemplate = require("lodash.template");
const _1 = require(".");
const Config = require("../config");
function extractExport(config, classPath) {
    const helpClassPath = Config.tsPath(config.root, classPath);
    return require(helpClassPath);
}
function extractClass(exported) {
    return exported && exported.default ? exported.default : exported;
}
function getHelpClass(config) {
    const pjson = config.pjson;
    const configuredClass = pjson && pjson.oclif && pjson.oclif.helpClass;
    if (configuredClass) {
        try {
            const exported = extractExport(config, configuredClass);
            return extractClass(exported);
        }
        catch (error) {
            throw new Error(`Unable to load configured help class "${configuredClass}", failed with message:\n${error.message}`);
        }
    }
    return _1.Help;
}
exports.getHelpClass = getHelpClass;
function template(context) {
    function render(t) {
        return lodashTemplate(t)(context);
    }
    return render;
}
exports.template = template;
function collateSpacedCmdIDFromArgs(argv, config) {
    if (argv.length === 1)
        return argv;
    const ids = config.commandIDs.concat(config.topics.map(t => t.name));
    const findId = (argv) => {
        const final = [];
        const idPresent = (id) => ids.includes(id);
        const isFlag = (s) => s.startsWith('-');
        const finalizeId = (s) => s ? [...final, s].join(':') : final.join(':');
        const hasSubCommandsWithArgs = () => {
            const subCommands = config.commands.filter(c => (c.id).startsWith(finalizeId()));
            return Boolean(subCommands.find(cmd => cmd.args.length > 0));
        };
        for (const arg of argv) {
            if (idPresent(finalizeId(arg)))
                final.push(arg);
            // If the parent topic has a command that expects positional arguments, then we cannot
            // assume that any subsequent string could be part of the command name
            else if (isFlag(arg) || hasSubCommandsWithArgs())
                break;
            else
                final.push(arg);
        }
        return finalizeId();
    };
    const id = findId(argv);
    if (id) {
        const argvSlice = argv.slice(id.split(':').length);
        return [id, ...argvSlice];
    }
    return argv; // ID is argv[0]
}
function toStandardizedId(commandID, config) {
    return commandID.replace(new RegExp(config.topicSeparator, 'g'), ':');
}
exports.toStandardizedId = toStandardizedId;
function toConfiguredId(commandID, config) {
    return commandID.replace(new RegExp(':', 'g'), config.topicSeparator);
}
exports.toConfiguredId = toConfiguredId;
function standarizeIDFromArgv(argv, config) {
    if (argv.length === 0)
        return argv;
    if (config.topicSeparator === ' ')
        argv = collateSpacedCmdIDFromArgs(argv, config);
    else if (config.topicSeparator !== ':')
        argv[0] = toStandardizedId(argv[0], config);
    return argv;
}
exports.standarizeIDFromArgv = standarizeIDFromArgv;
