// @lwc/template-compiler/dist/commonjs/parser/expression.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIteratorElement = exports.getForEachParent = exports.getForOfParent = exports.parseIdentifier = exports.parseExpression = exports.isPotentialExpression = exports.isExpression = exports.EXPRESSION_SYMBOL_END = exports.EXPRESSION_SYMBOL_START = void 0;
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const esutils = __importStar(require("esutils"));
const acorn_1 = require("acorn");
const errors_1 = require("@lwc/errors");
const t = __importStar(require("../shared/estree"));
exports.EXPRESSION_SYMBOL_START = '{';
exports.EXPRESSION_SYMBOL_END = '}';
const VALID_EXPRESSION_RE = /^{.+}$/;
const POTENTIAL_EXPRESSION_RE = /^.?{.+}.*$/;
const WHITESPACES_RE = /\s/;
function isExpression(source) {
    return !!source.match(VALID_EXPRESSION_RE);
}
exports.isExpression = isExpression;
function isPotentialExpression(source) {
    return !!source.match(POTENTIAL_EXPRESSION_RE);
}
exports.isPotentialExpression = isPotentialExpression;
function validateExpression(node, state) {
    const isValidNode = t.isIdentifier(node) || t.isMemberExpression(node);
    errors_1.invariant(isValidNode, errors_1.ParserDiagnostics.INVALID_NODE, [node.type]);
    if (t.isMemberExpression(node)) {
        errors_1.invariant(state.config.experimentalComputedMemberExpression || !node.computed, errors_1.ParserDiagnostics.COMPUTED_PROPERTY_ACCESS_NOT_ALLOWED);
        const { object, property } = node;
        if (!t.isIdentifier(object)) {
            validateExpression(object, state);
        }
        if (!t.isIdentifier(property)) {
            validateExpression(property, state);
        }
    }
}
function validateSourceIsParsedExpression(source, parsedExpression) {
    if (parsedExpression.end === source.length - 1) {
        return;
    }
    let unclosedParenthesisCount = 0;
    for (let i = 0, n = parsedExpression.start; i < n; i++) {
        if (source[i] === '(') {
            unclosedParenthesisCount++;
        }
    }
    // source[source.length - 1] === '}', n = source.length - 1 is to avoid processing '}'.
    for (let i = parsedExpression.end, n = source.length - 1; i < n; i++) {
        const character = source[i];
        if (character === ')') {
            unclosedParenthesisCount--;
        }
        else if (character === ';') {
            // acorn parseExpressionAt will stop at the first ";", it may be that the expression is not
            // a multiple expression ({foo;}), but this is a case that we explicitly do not want to support.
            // in such case, let's fail with the same error as if it were a multiple expression.
            errors_1.invariant(false, errors_1.ParserDiagnostics.MULTIPLE_EXPRESSIONS);
        }
        else {
            errors_1.invariant(WHITESPACES_RE.test(character), errors_1.ParserDiagnostics.TEMPLATE_EXPRESSION_PARSING_ERROR, ['Unexpected end of expression']);
        }
    }
    errors_1.invariant(unclosedParenthesisCount === 0, errors_1.ParserDiagnostics.TEMPLATE_EXPRESSION_PARSING_ERROR, [
        'Unexpected end of expression',
    ]);
}
function parseExpression(source, state) {
    try {
        const parsed = acorn_1.parseExpressionAt(source, 1, { ecmaVersion: 2020 });
        validateSourceIsParsedExpression(source, parsed);
        validateExpression(parsed, state);
        return parsed;
    }
    catch (err) {
        err.message = `Invalid expression ${source} - ${err.message}`;
        throw err;
    }
}
exports.parseExpression = parseExpression;
function parseIdentifier(source) {
    if (esutils.keyword.isIdentifierES6(source)) {
        return t.identifier(source);
    }
    else {
        throw errors_1.generateCompilerError(errors_1.ParserDiagnostics.INVALID_IDENTIFIER, {
            messageArgs: [source],
        });
    }
}
exports.parseIdentifier = parseIdentifier;
// Returns the immediate iterator parent if it exists.
// Traverses up until it finds an element with forOf, or
// a non-template element without a forOf.
function getForOfParent(element) {
    const parent = element.parent;
    if (!parent) {
        return null;
    }
    if (parent.forOf) {
        return parent;
    }
    else if (parent.tag.toLowerCase() === 'template') {
        return getForOfParent(parent);
    }
    return null;
}
exports.getForOfParent = getForOfParent;
function getForEachParent(element) {
    if (element.forEach) {
        return element;
    }
    const parent = element.parent;
    if (parent && parent.tag.toLowerCase() === 'template') {
        return getForEachParent(parent);
    }
    return null;
}
exports.getForEachParent = getForEachParent;
function isIteratorElement(element) {
    return !!(getForOfParent(element) || getForEachParent(element));
}
exports.isIteratorElement = isIteratorElement;
