// @oclif/core/lib/flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Parser = require("./parser");
function build(defaults) {
    return Parser.flags.build(defaults);
}
exports.build = build;
function option(options) {
    return build(options)();
}
exports.option = option;
const _enum = (opts) => {
    return build(Object.assign({ async parse(input) {
            if (!opts.options.includes(input))
                throw new Error(`Expected --${this.name}=${input} to be one of: ${opts.options.join(', ')}`);
            return input;
        }, helpValue: `(${opts.options.join('|')})` }, opts))();
};
exports.enum = _enum;
const stringFlag = build({});
exports.string = stringFlag;
var parser_1 = require("./parser");
exports.boolean = parser_1.boolean;
exports.integer = parser_1.integer;
exports.version = (opts = {}) => {
    return Parser.flags.boolean(Object.assign(Object.assign({ 
        // char: 'v',
        description: 'show CLI version' }, opts), { parse: async (_, cmd) => {
            cmd.log(cmd.config.userAgent);
            cmd.exit(0);
        } }));
};
exports.help = (opts = {}) => {
    return Parser.flags.boolean(Object.assign(Object.assign({ 
        // char: 'h',
        description: 'show CLI help' }, opts), { parse: async (_, cmd) => {
            cmd._help();
        } }));
};
